<center>

# **DSE302: Data Science and Machine Learning**
# **Final Project**


# **Identification of Exoplanets and Exomoons**

##### <font size="2">By:</font>

#### **Shaunak Padhyegurjar (19213)**
#### **Riya Rai (19234)**
#### **Yuvraj Sharma (19288)**
######  <font size="2"> *Department of Physics, IISER Bhopal* </font>

##### <font size="2">Instructor:</font>

#### Dr. Tanmay Basu
######  <font size="2"> *Assistant Professor, DSE Department, IISER Bhopal* </font>

<br>

<img src="https://dse.iiserb.ac.in/img/IISERB_Logo.png" alt="drawing" width="100"/>
</center>












# Author: Shaunak Padhyegurjar

# Code for Cleaning the training and testing data

## Importing the necessary packages

In [17]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

import sys
np.set_printoptions(threshold=sys.maxsize)

import warnings
warnings.filterwarnings("ignore")

## Importing the data

Here we import the training data and training labels for cleaning.

### Importing training X-data

In [18]:
df = pd.read_csv("/Users/shaunakpadhyegurjar/Downloads/Exoplanet data/exoplanet_trn_data.csv")
df

pl_name     hostname pl_letter    hd_name    hip_name  \
0           AU Mic c       AU Mic         c  HD 197481  HIP 102409   
1      Kepler-1476 b  Kepler-1476         b        NaN         NaN   
2       Kepler-255 d   Kepler-255         d        NaN         NaN   
3        Kepler-82 e    Kepler-82         e        NaN         NaN   
4           K2-156 b       K2-156         b        NaN         NaN   
...              ...          ...       ...        ...         ...   
17964   Kepler-245 d   Kepler-245         d        NaN         NaN   
17965    Kepler-28 b    Kepler-28         b        NaN         NaN   
17966  Kepler-1272 b  Kepler-1272         b        NaN         NaN   
17967    HD 102117 b    HD 102117         b  HD 102117   HIP 57291   
17968    Kepler-58 c    Kepler-58         c        NaN         NaN   

              tic_id                       gaia_id  default_flag  sy_snum  \
0      TIC 441420236  Gaia DR2 6794047652729201024             1        1   
1      TIC 164727404  Gaia DR2 2106597487787652352             0        1   
2      TIC 272083669  Gaia DR2 2080202611655694336             0        1   
3       TIC 63372898  Gaia DR2 2125850623586710400             0        1   
4      TIC 176868262  Gaia DR2 3596250888028092160             1        1   
...              ...                           ...           ...      ...   
17964   TIC 63004906  Gaia DR2 2101768089124657664             0        1   
17965   TIC 63122956  Gaia DR2 2125731464014361088             1        1   
17966  TIC 399794186  Gaia DR2 2100148611573075200             0        1   
17967  TIC 267519701  Gaia DR2 5341673851394297600             0        1   
17968  TIC 184471242  Gaia DR2 2073274618947678464             0        1   

       sy_pnum  ...  st_nrvc  st_nspec pl_nespec  pl_ntranspec Unnamed: 282  \
0            2  ...        0         1         0             0          NaN   
1            1  ...        0         0         0             0          NaN   
2            3  ...        0         0         0             0          NaN   
3            5  ...        0         0         0             0          NaN   
4            1  ...        0         0         0             0          NaN   
...        ...  ...      ...       ...       ...           ...          ...   
17964        4  ...        0         0         0             0          NaN   
17965        2  ...        0         0         0             0          NaN   
17966        1  ...        0         0         0             0          0.0   
17967        1  ...        3         0         0             0          NaN   
17968        3  ...        0         0         0             0          NaN   

      Unnamed: 283 Unnamed: 284 Unnamed: 285 Unnamed: 286 Unnamed: 287  
0              NaN          NaN          NaN          NaN          NaN  
1              NaN          NaN          NaN          NaN          NaN  
2              NaN          NaN          NaN          NaN          NaN  
3              NaN          NaN          NaN          NaN          NaN  
4              NaN          NaN          NaN          NaN          NaN  
...            ...          ...          ...          ...          ...  
17964          NaN          NaN          NaN          NaN          NaN  
17965          NaN          NaN          NaN          NaN          NaN  
17966          NaN          NaN          NaN          NaN          NaN  
17967          NaN          NaN          NaN          NaN          NaN  
17968          NaN          NaN          NaN          NaN          NaN  

[17969 rows x 288 columns]

### Importing training labels (y-data)

In [19]:
labels = pd.read_csv('/Users/shaunakpadhyegurjar/Downloads/Exoplanet data/exoplanet_trn_data_targets.csv', header=None, usecols=[1])
labels

1
0      0.1101
1      0.0932
2      0.0202
3      0.0632
4      0.0149
...       ...
17964  0.1995
17965  0.0620
17966  0.0000
17967  0.1532
17968  0.1228

[17969 rows x 1 columns]

## Data Cleaning

The data contains some rows in which the entries of a column appear in the next column. Moreover, some rows are such that the values are shifted by multiple spaces down the row. We observed that for such rows, the value of target variable was 0 meaning it was non existent. There was no point in correcting the shift in these rows as the corresponding value of target variable being zero will decrease the performance of our models. Hence we removed these faulty rows. 

The last few columns of the training data (X-data) are unnamed. We rename those columns for easy referencing. These last columns are useful in detecting faulty rows. 

In [20]:
df.rename(columns = {'Unnamed: 282':'shift1', 'Unnamed: 283':'shift2', 'Unnamed: 284':'shift3', 
                     'Unnamed: 285':'shift4', 'Unnamed: 286':'shift5', 'Unnamed: 287':'shift6'}, inplace = True)
df

pl_name     hostname pl_letter    hd_name    hip_name  \
0           AU Mic c       AU Mic         c  HD 197481  HIP 102409   
1      Kepler-1476 b  Kepler-1476         b        NaN         NaN   
2       Kepler-255 d   Kepler-255         d        NaN         NaN   
3        Kepler-82 e    Kepler-82         e        NaN         NaN   
4           K2-156 b       K2-156         b        NaN         NaN   
...              ...          ...       ...        ...         ...   
17964   Kepler-245 d   Kepler-245         d        NaN         NaN   
17965    Kepler-28 b    Kepler-28         b        NaN         NaN   
17966  Kepler-1272 b  Kepler-1272         b        NaN         NaN   
17967    HD 102117 b    HD 102117         b  HD 102117   HIP 57291   
17968    Kepler-58 c    Kepler-58         c        NaN         NaN   

              tic_id                       gaia_id  default_flag  sy_snum  \
0      TIC 441420236  Gaia DR2 6794047652729201024             1        1   
1      TIC 164727404  Gaia DR2 2106597487787652352             0        1   
2      TIC 272083669  Gaia DR2 2080202611655694336             0        1   
3       TIC 63372898  Gaia DR2 2125850623586710400             0        1   
4      TIC 176868262  Gaia DR2 3596250888028092160             1        1   
...              ...                           ...           ...      ...   
17964   TIC 63004906  Gaia DR2 2101768089124657664             0        1   
17965   TIC 63122956  Gaia DR2 2125731464014361088             1        1   
17966  TIC 399794186  Gaia DR2 2100148611573075200             0        1   
17967  TIC 267519701  Gaia DR2 5341673851394297600             0        1   
17968  TIC 184471242  Gaia DR2 2073274618947678464             0        1   

       sy_pnum  ...  st_nrvc  st_nspec pl_nespec  pl_ntranspec shift1 shift2  \
0            2  ...        0         1         0             0    NaN    NaN   
1            1  ...        0         0         0             0    NaN    NaN   
2            3  ...        0         0         0             0    NaN    NaN   
3            5  ...        0         0         0             0    NaN    NaN   
4            1  ...        0         0         0             0    NaN    NaN   
...        ...  ...      ...       ...       ...           ...    ...    ...   
17964        4  ...        0         0         0             0    NaN    NaN   
17965        2  ...        0         0         0             0    NaN    NaN   
17966        1  ...        0         0         0             0    0.0    NaN   
17967        1  ...        3         0         0             0    NaN    NaN   
17968        3  ...        0         0         0             0    NaN    NaN   

      shift3 shift4 shift5 shift6  
0        NaN    NaN    NaN    NaN  
1        NaN    NaN    NaN    NaN  
2        NaN    NaN    NaN    NaN  
3        NaN    NaN    NaN    NaN  
4        NaN    NaN    NaN    NaN  
...      ...    ...    ...    ...  
17964    NaN    NaN    NaN    NaN  
17965    NaN    NaN    NaN    NaN  
17966    NaN    NaN    NaN    NaN  
17967    NaN    NaN    NaN    NaN  
17968    NaN    NaN    NaN    NaN  

[17969 rows x 288 columns]

Now we count the number of rows for which the value of target variable is zero.

In [21]:
index1 = np.where(np.array(labels) == 0)[0]
len(index1)

2076

We see that there are 2076 rows for which the target variable is zero. Upon inspecting the data manually, we found that the `pl_orbper` (planet orbital period) column had several instances of strings where there should be float values. Actually, most of these strings lie in the same rows for which the target variable is zero. Hence we now check if the count of the rows containing strings and count of rows for which target variable is zero (which is 2076) are same or not.

In [22]:
period = np.array(df["pl_orbper"])
index2 = np.array([])
count = 0
for i in range(len(period)):
    try:
        float(period[i])
    except:
        count += 1
        index2 = np.append(index2, i)
    period[i] = str(period[i]) 
        
count

2185

We see that the number of rows containing strings are more than the number of rows for which target variable is zero. Hence we remove the rows containing strings from the dataset as this would also remove the rows for which target variable is zero. Even though this means some rows for which target variable is non-zero will also be removed, this doesn't cause much loss of information/datapoints as the two counts are quite close.

In [23]:
df.drop(index2, axis=0, inplace=True)
df.reset_index()
df

pl_name     hostname pl_letter    hd_name    hip_name  \
0           AU Mic c       AU Mic         c  HD 197481  HIP 102409   
1      Kepler-1476 b  Kepler-1476         b        NaN         NaN   
2       Kepler-255 d   Kepler-255         d        NaN         NaN   
3        Kepler-82 e    Kepler-82         e        NaN         NaN   
4           K2-156 b       K2-156         b        NaN         NaN   
...              ...          ...       ...        ...         ...   
17963   Kepler-138 c   Kepler-138         c        NaN         NaN   
17964   Kepler-245 d   Kepler-245         d        NaN         NaN   
17965    Kepler-28 b    Kepler-28         b        NaN         NaN   
17967    HD 102117 b    HD 102117         b  HD 102117   HIP 57291   
17968    Kepler-58 c    Kepler-58         c        NaN         NaN   

              tic_id                       gaia_id  default_flag  sy_snum  \
0      TIC 441420236  Gaia DR2 6794047652729201024             1        1   
1      TIC 164727404  Gaia DR2 2106597487787652352             0        1   
2      TIC 272083669  Gaia DR2 2080202611655694336             0        1   
3       TIC 63372898  Gaia DR2 2125850623586710400             0        1   
4      TIC 176868262  Gaia DR2 3596250888028092160             1        1   
...              ...                           ...           ...      ...   
17963  TIC 159376971  Gaia DR2 2102053446751282944             0        1   
17964   TIC 63004906  Gaia DR2 2101768089124657664             0        1   
17965   TIC 63122956  Gaia DR2 2125731464014361088             1        1   
17967  TIC 267519701  Gaia DR2 5341673851394297600             0        1   
17968  TIC 184471242  Gaia DR2 2073274618947678464             0        1   

       sy_pnum  ...  st_nrvc  st_nspec pl_nespec  pl_ntranspec shift1 shift2  \
0            2  ...        0         1         0             0    NaN    NaN   
1            1  ...        0         0         0             0    NaN    NaN   
2            3  ...        0         0         0             0    NaN    NaN   
3            5  ...        0         0         0             0    NaN    NaN   
4            1  ...        0         0         0             0    NaN    NaN   
...        ...  ...      ...       ...       ...           ...    ...    ...   
17963        3  ...        0         0         0             2    NaN    NaN   
17964        4  ...        0         0         0             0    NaN    NaN   
17965        2  ...        0         0         0             0    NaN    NaN   
17967        1  ...        3         0         0             0    NaN    NaN   
17968        3  ...        0         0         0             0    NaN    NaN   

      shift3 shift4 shift5 shift6  
0        NaN    NaN    NaN    NaN  
1        NaN    NaN    NaN    NaN  
2        NaN    NaN    NaN    NaN  
3        NaN    NaN    NaN    NaN  
4        NaN    NaN    NaN    NaN  
...      ...    ...    ...    ...  
17963    NaN    NaN    NaN    NaN  
17964    NaN    NaN    NaN    NaN  
17965    NaN    NaN    NaN    NaN  
17967    NaN    NaN    NaN    NaN  
17968    NaN    NaN    NaN    NaN  

[15784 rows x 288 columns]

**We drop the same rows from the target variable column too:**

In [24]:
labels.drop(index2, axis=0, inplace=True)

So far we have removed most of the faulty rows but not all. Upon manual inspection of the resulting data, we found that some faulty rows were still remaining which didn't satisfy the previous two conditions. Now we use the previouly mentioned unnamed columns in the data to detect these faulty rows and delete them.

In [25]:
nonzero = np.array([0.0, 1.0, 2.0, 3.0, 4.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 32.0, 36.0, 45.0, 69.0, 
                    89.0, 40.0, 91.0, 42.0, 47.0, 111.0])

shift1 = np.array(df["shift1"])
index = np.array(df.index)


delete_rows = np.array([])

for i in range(len(shift1)):
    if shift1[i] in nonzero:
        delete_rows = np.append(delete_rows, index[i])
        
delete_rows

array([  884.,  1285.,  2180.,  2251.,  4396.,  8276., 10210., 11816.,
       15255.])

In [26]:
df.drop(delete_rows, inplace=True, axis=0)
df

pl_name     hostname pl_letter    hd_name    hip_name  \
0           AU Mic c       AU Mic         c  HD 197481  HIP 102409   
1      Kepler-1476 b  Kepler-1476         b        NaN         NaN   
2       Kepler-255 d   Kepler-255         d        NaN         NaN   
3        Kepler-82 e    Kepler-82         e        NaN         NaN   
4           K2-156 b       K2-156         b        NaN         NaN   
...              ...          ...       ...        ...         ...   
17963   Kepler-138 c   Kepler-138         c        NaN         NaN   
17964   Kepler-245 d   Kepler-245         d        NaN         NaN   
17965    Kepler-28 b    Kepler-28         b        NaN         NaN   
17967    HD 102117 b    HD 102117         b  HD 102117   HIP 57291   
17968    Kepler-58 c    Kepler-58         c        NaN         NaN   

              tic_id                       gaia_id  default_flag  sy_snum  \
0      TIC 441420236  Gaia DR2 6794047652729201024             1        1   
1      TIC 164727404  Gaia DR2 2106597487787652352             0        1   
2      TIC 272083669  Gaia DR2 2080202611655694336             0        1   
3       TIC 63372898  Gaia DR2 2125850623586710400             0        1   
4      TIC 176868262  Gaia DR2 3596250888028092160             1        1   
...              ...                           ...           ...      ...   
17963  TIC 159376971  Gaia DR2 2102053446751282944             0        1   
17964   TIC 63004906  Gaia DR2 2101768089124657664             0        1   
17965   TIC 63122956  Gaia DR2 2125731464014361088             1        1   
17967  TIC 267519701  Gaia DR2 5341673851394297600             0        1   
17968  TIC 184471242  Gaia DR2 2073274618947678464             0        1   

       sy_pnum  ...  st_nrvc  st_nspec pl_nespec  pl_ntranspec shift1 shift2  \
0            2  ...        0         1         0             0    NaN    NaN   
1            1  ...        0         0         0             0    NaN    NaN   
2            3  ...        0         0         0             0    NaN    NaN   
3            5  ...        0         0         0             0    NaN    NaN   
4            1  ...        0         0         0             0    NaN    NaN   
...        ...  ...      ...       ...       ...           ...    ...    ...   
17963        3  ...        0         0         0             2    NaN    NaN   
17964        4  ...        0         0         0             0    NaN    NaN   
17965        2  ...        0         0         0             0    NaN    NaN   
17967        1  ...        3         0         0             0    NaN    NaN   
17968        3  ...        0         0         0             0    NaN    NaN   

      shift3 shift4 shift5 shift6  
0        NaN    NaN    NaN    NaN  
1        NaN    NaN    NaN    NaN  
2        NaN    NaN    NaN    NaN  
3        NaN    NaN    NaN    NaN  
4        NaN    NaN    NaN    NaN  
...      ...    ...    ...    ...  
17963    NaN    NaN    NaN    NaN  
17964    NaN    NaN    NaN    NaN  
17965    NaN    NaN    NaN    NaN  
17967    NaN    NaN    NaN    NaN  
17968    NaN    NaN    NaN    NaN  

[15775 rows x 288 columns]

In [27]:
labels.drop(delete_rows, axis=0, inplace=True)
labels

1
0      0.1101
1      0.0932
2      0.0202
3      0.0632
4      0.0149
...       ...
17963  0.0905
17964  0.1995
17965  0.0620
17967  0.1532
17968  0.1228

[15775 rows x 1 columns]

We now have the cleaned data ready which we will save to new `.csv` files for further use in running the models.

In [28]:
df.to_csv("/Users/shaunakpadhyegurjar/Downloads/Exoplanet data/cleaned train data.csv", index=None)
labels.to_csv("/Users/shaunakpadhyegurjar/Downloads/Exoplanet data/cleaned train labels.csv", index=None)

## Cleaning the Testing data

We also need to clean the testing data as it too contains strings and faulty rows. In this case, deleting the rows with strings was sufficient.

In [29]:
test_df = pd.read_csv("/Users/shaunakpadhyegurjar/Downloads/Exoplanet data/exoplanet_tst_data.csv")
test_df

pl_name    hostname pl_letter   hd_name   hip_name         tic_id  \
0       HD 98219 b    HD 98219         b  HD 98219  HIP 55174  TIC 437246853   
1     Kepler-302 b  Kepler-302         b       NaN        NaN  TIC 270951852   
2      Kepler-63 b   Kepler-63         b       NaN        NaN  TIC 299158887   
3     Kepler-778 b  Kepler-778         b       NaN        NaN  TIC 158484518   
4     Kepler-757 b  Kepler-757         b       NaN        NaN  TIC 360022783   
...            ...         ...       ...       ...        ...            ...   
1992  Kepler-415 c  Kepler-415         c       NaN        NaN  TIC 138365520   
1993  Kepler-423 b  Kepler-423         b       NaN        NaN   TIC 63452790   
1994  Kepler-825 c  Kepler-825         c       NaN        NaN  TIC 378086008   
1995  Kepler-688 b  Kepler-688         b       NaN        NaN  TIC 378085713   
1996  Kepler-155 c  Kepler-155         c       NaN        NaN  TIC 298969755   

                           gaia_id  default_flag  sy_snum  sy_pnum  ...  \
0     Gaia DR2 3534621989981179520             1        1        1  ...   
1     Gaia DR2 2078195212660287616             0        1        2  ...   
2     Gaia DR2 2132628489996257920             0        1        1  ...   
3     Gaia DR2 2106046632461881856             0        1        1  ...   
4     Gaia DR2 2076848310922066048             0        1        1  ...   
...                            ...           ...      ...      ...  ...   
1992  Gaia DR2 2052280548230453760             0        1        2  ...   
1993  Gaia DR2 2128273225659492608             0        1        1  ...   
1994  Gaia DR2 2076249351963553152             0        1        2  ...   
1995  Gaia DR2 2052232410241029248             0        1        1  ...   
1996  Gaia DR2 2133351698066484096             0        1        2  ...   

      st_nrvc  st_nspec pl_nespec  pl_ntranspec Unnamed: 282 Unnamed: 283  \
0           0         1         0             0          NaN          NaN   
1           0         0         0             0          NaN          NaN   
2           0         0         0             0          NaN          NaN   
3           0         0         0             0          NaN          NaN   
4           0         0         0             0          NaN          NaN   
...       ...       ...       ...           ...          ...          ...   
1992        0         0         0             0          NaN          NaN   
1993        0         0         0             0          NaN          NaN   
1994        0         0         0             0          NaN          NaN   
1995        0         0         0             0          NaN          NaN   
1996        0         0         0             0          NaN          NaN   

     Unnamed: 284 Unnamed: 285 Unnamed: 286 Unnamed: 287  
0             NaN          NaN          NaN          NaN  
1             NaN          NaN          NaN          NaN  
2             NaN          NaN          NaN          NaN  
3             NaN          NaN          NaN          NaN  
4             NaN          NaN          NaN          NaN  
...           ...          ...          ...          ...  
1992          NaN          NaN          NaN          NaN  
1993          NaN          NaN          NaN          NaN  
1994          NaN          NaN          NaN          NaN  
1995          NaN          NaN          NaN          NaN  
1996          NaN          NaN          NaN          NaN  

[1997 rows x 288 columns]

In [30]:
period = np.array(test_df["pl_orbper"])
index2 = np.array([])
count = 0
for i in range(len(period)):
    try:
        float(period[i])
    except:
        count += 1
        index2 = np.append(index2, i)
    period[i] = str(period[i]) 
        
count

224

There are 224 faulty rows in the testing data which we remove.

In [31]:
test_df.drop(index2, axis=0, inplace=True)
test_df.reset_index()
test_df

pl_name    hostname pl_letter   hd_name   hip_name         tic_id  \
0       HD 98219 b    HD 98219         b  HD 98219  HIP 55174  TIC 437246853   
1     Kepler-302 b  Kepler-302         b       NaN        NaN  TIC 270951852   
2      Kepler-63 b   Kepler-63         b       NaN        NaN  TIC 299158887   
3     Kepler-778 b  Kepler-778         b       NaN        NaN  TIC 158484518   
4     Kepler-757 b  Kepler-757         b       NaN        NaN  TIC 360022783   
...            ...         ...       ...       ...        ...            ...   
1992  Kepler-415 c  Kepler-415         c       NaN        NaN  TIC 138365520   
1993  Kepler-423 b  Kepler-423         b       NaN        NaN   TIC 63452790   
1994  Kepler-825 c  Kepler-825         c       NaN        NaN  TIC 378086008   
1995  Kepler-688 b  Kepler-688         b       NaN        NaN  TIC 378085713   
1996  Kepler-155 c  Kepler-155         c       NaN        NaN  TIC 298969755   

                           gaia_id  default_flag  sy_snum  sy_pnum  ...  \
0     Gaia DR2 3534621989981179520             1        1        1  ...   
1     Gaia DR2 2078195212660287616             0        1        2  ...   
2     Gaia DR2 2132628489996257920             0        1        1  ...   
3     Gaia DR2 2106046632461881856             0        1        1  ...   
4     Gaia DR2 2076848310922066048             0        1        1  ...   
...                            ...           ...      ...      ...  ...   
1992  Gaia DR2 2052280548230453760             0        1        2  ...   
1993  Gaia DR2 2128273225659492608             0        1        1  ...   
1994  Gaia DR2 2076249351963553152             0        1        2  ...   
1995  Gaia DR2 2052232410241029248             0        1        1  ...   
1996  Gaia DR2 2133351698066484096             0        1        2  ...   

      st_nrvc  st_nspec pl_nespec  pl_ntranspec Unnamed: 282 Unnamed: 283  \
0           0         1         0             0          NaN          NaN   
1           0         0         0             0          NaN          NaN   
2           0         0         0             0          NaN          NaN   
3           0         0         0             0          NaN          NaN   
4           0         0         0             0          NaN          NaN   
...       ...       ...       ...           ...          ...          ...   
1992        0         0         0             0          NaN          NaN   
1993        0         0         0             0          NaN          NaN   
1994        0         0         0             0          NaN          NaN   
1995        0         0         0             0          NaN          NaN   
1996        0         0         0             0          NaN          NaN   

     Unnamed: 284 Unnamed: 285 Unnamed: 286 Unnamed: 287  
0             NaN          NaN          NaN          NaN  
1             NaN          NaN          NaN          NaN  
2             NaN          NaN          NaN          NaN  
3             NaN          NaN          NaN          NaN  
4             NaN          NaN          NaN          NaN  
...           ...          ...          ...          ...  
1992          NaN          NaN          NaN          NaN  
1993          NaN          NaN          NaN          NaN  
1994          NaN          NaN          NaN          NaN  
1995          NaN          NaN          NaN          NaN  
1996          NaN          NaN          NaN          NaN  

[1773 rows x 288 columns]

Finally we save the cleaned testing data to new `.csv` file too.

In [32]:
test_df.to_csv("/Users/shaunakpadhyegurjar/Downloads/Exoplanet data/cleaned test data.csv", index=None)